File: apps/mobile/packages/bvc_network/lib/src/api_response.dart
class ApiResponse<T> {
  ApiResponse({
    required this.success,
    required this.data,
    required this.message,
    required this.meta,
  });

  final bool success;
  final T data;
  final String message;
  final Map<String, dynamic> meta;
}

ApiResponse<T> parseApiResponse<T>(
  Map<String, dynamic> json,
  T Function(dynamic data) parseData,
) {
  final success = json['success'] == true;
  final message = (json['message'] as String?) ?? '';
  final meta = (json['meta'] as Map?)?.cast<String, dynamic>() ?? <String, dynamic>{};

  if (!success) {
    throw Exception(message.isNotEmpty ? message : 'Request failed');
  }

  return ApiResponse<T>(
    success: true,
    data: parseData(json['data']),
    message: message,
    meta: meta,
  );
}

File: apps/api/src/common/response/api-response.ts
export type ApiResponse<T> = {
  success: boolean;
  data: T;
  message: string;
  meta?: Record<string, unknown>;
};

export function apiSuccess<T>(
  data: T,
  message = 'OK',
  meta?: Record<string, unknown>,
): ApiResponse<T> {
  return { success: true, data, message, meta };
}

export function apiError(
  message = 'Error',
  meta?: Record<string, unknown>,
): ApiResponse<null> {
  return { success: false, data: null, message, meta };
}
